### PR TITLE
MAINT: Added clang-format formatting for special C++ files

### PR DESCRIPTION
**Pull request by izaid**
#### Reference issue

https://github.com/scipy/scipy/pull/19601

#### What does this implement/fix?

This adds simple autoformatting for the C++ special files, as discussed in the above PR. It is just clang-format, a widely used tool for C++ code, the standard of which is specified by the added .clang-format file. This only affects the new C++ files in special.

Of course, we can change the formatting standard as we go, it is just nice to have something.

### CODE REVIEW

File: scipy/special/special/error.h [hunk 1]
```diff
@@ -12,31 +13,30 @@ extern "C" {
...
+  }
 #else
-        void set_error(const char *func_name, sf_error_t code, const char *fmt, ...);
+  void set_error(const char *func_name, sf_error_t code, const char *fmt, ...);
 #endif
-}
+} // namespace special
 
 } // closes extern "C"
 #endif
```
Comment (lucascolley) [inline scipy/special/special/error.h:42]:
```suggestion
#endif

```

File: scipy/special/special/evalpoly.h [hunk 2]
```diff
@@ -18,26 +18,25 @@
...
+  }
 
-	return z*a + b;
-    }
+  return z * a + b;
+}
 
-}
\ No newline at end of file
+} // namespace special
```
Comment (lucascolley) [inline scipy/special/special/evalpoly.h:42]:
```suggestion
} // namespace special

```

File: scipy/special/special/lambertw.h [hunk 3]
```diff
@@ -10,144 +10,129 @@
...
+    }
+  }
+
+  set_error("lambertw", SF_ERROR_SLOW, "iteration failed to converge: %g + %gj", z.real(), z.imag());
+  return std::complex<double>(std::numeric_limits<double>::quiet_NaN(), std::numeric_limits<double>::quiet_NaN());
+}
 
-}
\ No newline at end of file
+} // namespace special
```
Comment (lucascolley) [inline scipy/special/special/lambertw.h:138]:
```suggestion
} // namespace special

```

**Review by lucascolley — COMMENTED** [review]
not that it matters with modern C++, but may as well have newlines at the end of files while we're on the topic of style

**Comment by izaid** [comment]
I'm happy with newlines at the end of each file, and there is an option for this in clang-format (InsertNewlineAtEOF). But it is not supported until version 16! I'm on Ubuntu 22.04, and only have version 14.

**Comment by lucascolley** [comment]
> there is an option for this in clang-format (InsertNewlineAtEOF). But it is not supported until version 16

Will the formatter remove new lines with the current configuration? If so, then no worries. If not, then we can push the manual changes.

**Comment by izaid** [comment]
@lucascolley Seems like it doesn't touch the newlines if we put them there manually. So have done so!

**Review by steppi — APPROVED** [review]
Thanks Irwin! The style looks much better now.

**Comment by ilayn** [comment]
Could we please stick to 4 spaces? This is extremely difficult also on Fortran files, it's not an urgency but seriously difficult when you are going around like a madman inside very long if blocks. Also we are the python folks afterall

**Comment by izaid** [comment]
> Could we please stick to 4 spaces? This is extremely difficult also on Fortran files, it's not an urgency but seriously difficult when you are going around like a madman inside very long if blocks. Also we are the python folks afterall

Happy with this, will change it when I get a chance.

**Comment by steppi** [comment]
Sorry, pulled the trigger too soon merging. Yeah, 4 spaces is fine.

**Comment by rgommers** [comment]
In addition, in the next PR, can you please add entries for style-only commits in `.git-blame-ignore-revs`? We want to keep `git blame` as useful as possible.

+1 for adding clang-format support though, thanks for doing that.

**Pull request merged**